### DISCUSSION TITLE
Producer Reference Time - AVPacket side_data_list - frame wallclock

### DISCUSSION THREAD

**Discussion by AvengerIl** [bluenviron/mediamtx#1134] [Questions]
Im quite stumped, I've been hacking FFMpeg and Gstreamer to their core but can only reliably hack ffprobe to show me this data, an ip camera stamps each frame with an NTP time, I can see it in ffmpeg/ffprobe (with mods) but cannot forward it in the stream.

I believe rtsp-simple-serv does something with this clock, I believe its aware of it, ideally I would like to modify it so it takes it from packet side_data into frame user data, somewhere where gstreamer can decode it and possibly save it along with the frames.

Im new to go, and so far I havent found evidence of r-s-s actually handling this data.... Would very much appreciate any help.

h264.

(fffmpeg)libavcodec/avpacket.c: 415: AV_PKT_DATA_PRFT

ffprobe -show_packets -of json rtsp://<ip>:<port>/proxied

"side_data_list": [
    {
        "side_data_type": "Producer Reference Time",
        "wallclock": 1662426900261104,
        "flags": 24
    }

**Comment by aler9** [comment]
Hello, NTP timestamp of each frame is already routed by RTSP by using RTCP and RTP packets.

RTCP packets are sent periodically to readers (once every 10 seconds) and contains the number of microseconds from 1st january 1970 and a relative timestamp (uint32) that corresponds to that NTP. Before v0.20.1 they were sent to readers only if the transport protocol was UDP, now they're also sent if the transport protocol is TCP.

RTP packets contain frames (encoded with H264) and a relative timestamp (uint32); therefore, their NTP timestamp is

```
ntp_timestamp_of_last_rtcp_packet + (rtp_packet_relative_timestamp - relative_timestamp_of_last_rtcp_packet) * clock_rate
```

FFmpeg already does the computation, and you can access the NTP timestamp by using `start_time_realtime` of `AVFormatContext`

[a comment by an automated account is not shown]
